iocage_lib/ioc_plugin.py

```diff
@@ -84,6 +87,148 @@ def __init__(self, release=None, jail=None, plugin=None, branch=None,
+        def download_parse_packagesite(packagesite_url):
+            package_site_data = {}
+            try:
+                with requests.get(
+                    f'{packagesite_url}/packagesite.txz',
+                    stream=True
+                ) as r:
+                    r.raise_for_status()
+                    with tempfile.NamedTemporaryFile(
+                        'wb+', delete=False
+                    ) as f:
+                        packagesite_path = f.name
+                        for chunk in r.iter_content(chunk_size=8192):
+                            if chunk:
+                                f.write(chunk)
+            except Exception:
+                pass
+            else:
+                with tempfile.TemporaryDirectory() as td:
+                    tar = tarfile.open(packagesite_path)
+                    tar.extractall(path=td)
+                    tar.close()
+                    with iocage_lib.ioc_exceptions.ignore_exceptions(
+                        Exception
+                    ):
+                        os.remove(packagesite_path)
+
+                    with open(
+                        os.path.join(td, 'packagesite.yaml'), 'rb'
+                    ) as f:
+                        for line in f.readlines():
+                            try:
+                                line_data = json.loads(line.decode())
+                            except UnicodeDecodeError:
+                                pass
+                            else:
+                                package_site_data[
+                                    line_data['origin']
+                                ] = line_data
+
+            return packagesite_url, package_site_data
+
+        plugin_packagesite_mapping = {}
+        package_sites = set([
+            url[:-1] if url[-1] == '/' else url for url in package_sites
+        ])
+
+        with concurrent.futures.ThreadPoolExecutor() as exc:
+            futures = exc.map(
+                download_parse_packagesite, package_sites
+            )
+
+            for future in futures:
+                plugin_packagesite_mapping[future[0]] = future[1]
+
+        return plugin_packagesite_mapping
+
+    @staticmethod
+    def fetch_plugin_versions_from_plugin_index(plugins_index):
+        plugin_packagesite_mapping = IOCPlugin.fetch_plugin_packagesites([
+            v['packagesite'] for v in plugins_index.values()
+            if v.get('packagesite')
+        ])
+
+        version_dict = {}
+        for plugin in plugins_index:
+            plugin_dict = plugins_index[plugin]
+            packagesite = plugin_dict.get('packagesite')
+            primary_package = plugin_dict.get('primary_pkg') or plugin
+            if packagesite:
+                packagesite = packagesite[:-1] if packagesite[-1] == '/' \
+                    else packagesite
+                plugin_pkgs = plugin_packagesite_mapping[packagesite]
+                for pkg in plugin_pkgs:
```

Comment:
So we are iterating through every single package in each packagesite? Did I get this right?

That does not seem very good (O(n)). We should come up with a way to make it O(log n) using a better structure.

This also likely means using two indexes because origin and package name, see the other discussion about origin.